framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,552 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <tinyara/audio/audio.h>
+#include <tinyara/audio/pcm.h>
+#include <tinyalsa/tinyalsa.h>
+#include <tinyara/fs/ioctl.h>
+#include <pthread.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <sched.h>
+#include <string.h>
+#include <mediaplayer/mediaplayer.h>
+#include "mplayer_debug.h"
+
+/***
+ * Structure
+ */
+struct media_s {
+	dq_entry_t q;
+	int fd;
+	media_op_t op;
+	media_state_t state;
+	media_type_t type;
+	media_play_type_t play_type;
+	struct pcm_config *config;
+};
+
+/***
+ * Static variables
+ */
+static pthread_t g_pth_playing;
+#ifdef CONFIG_AUDIO_MULTI_CARD
+static pthread_t g_pth_recording;
+#endif
+static sem_t g_sem_playing;
+static sem_t g_sem_recording;
+static pthread_mutex_t g_mutex_playing_q;
+static pthread_mutex_t g_mutex_recording_q;
+
+static bool g_playing_live;
+static bool g_recording_live;
+
+static dq_queue_t g_playing_q;
+static dq_queue_t g_recording_q;
+
+/***
+ * Pre-definitions
+ */
+media_result_t audio_playing(void *args);
+media_result_t audio_recording(void *args);
+
+media_result_t media_init(struct media_cb_s *cbs)
+{
+	pthread_attr_t attr;
+	pthread_mutexattr_t mutexAttr;
+	struct sched_param sparam;
+
+	// Init dq and sem
+	dq_init(&g_playing_q);
+	sem_init(&g_sem_playing, 0, 0);
+
+	pthread_mutexattr_init(&mutexAttr);
+	pthread_mutexattr_settype(&mutexAttr, PTHREAD_MUTEX_NORMAL);
+	pthread_mutex_init(&g_mutex_playing_q, &mutexAttr);
+
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	dq_init(&g_recording_q);
+	sem_init(&g_sem_recording, 0, 0);
+	pthread_mutex_init(&g_mutex_recording_q, &mutexAttr);
+#endif
+	// Init threads
+	if (pthread_attr_init(&attr) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	sparam.sched_priority = 150;
+	if (pthread_attr_setschedparam(&attr, &sparam) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setschedpolicy(&attr, SCHED_RR) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_attr_setstacksize(&attr, 8192) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+
+	if (pthread_create(&g_pth_playing, &attr, (pthread_startroutine_t)audio_playing, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
+	}
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	if (pthread_create(&g_pth_recording, &attr, (pthread_startroutine_t)audio_recording, (void *)NULL) != 0) {
+		return MEDIA_ERROR_THREAD_INIT;
```

Comment:
When making recording thread is failed, don't we need to cancel previous thread?

Reply:
There are many assets apart from pthread cancellation which should be destroyed.
Actually , I thought of usage of mediaplayer like this ;-
if (media_init() == MEDIA_ERROR){
         media_shutdown()
}

If this is not fine, then I will change it.

Reply:
Definitely not fine.
```
if (open == fail) {
  close
}
```
Do you think is it fine?
```
if (open == fail) {
  retry or return
}
```
I think this is a normal case.